### ISSUE TITLE
XConn methods should return errors and be fully documented

### ISSUE THREAD

**Issue by sminez** [sminez/penrose#122]
Title says it all: there are lots of things that can error when communicating with the X server and these are all silently dropped by the current xcb back end. The `*_checked` methods should be used where possible and `XConn` should be updated to return Results in cases where operations can fail.

**Comment by psychon** [comment]
> these are all silently dropped by the current xcb back end.

I was going to say "what? nooo, that's not how libxcb works! here, I'll show you an example". Then I spent about an error trying to figure out why `self.mark_focused_window(0xdeadbeef);` doesn't show up as an X11 error (event with response type 0). In the end, I figured out that the XCB crate is... I don't know what to call it.

I reported https://github.com/rtbo/rust-xcb/issues/97 about this.

That's the third rust-xcb issue that I opened due to penrose. Yesterday, I already reported https://github.com/rtbo/rust-xcb/issues/95 and https://github.com/rtbo/rust-xcb/issues/96. Sigh. :cry: 

Anyway, this is what I tried (and this also contains what I believe to be the fix for #124, but I am confused why the code works for you; the code definitely mistakes X11 errors for `RandR::ScreenChangeNotify` events):
```diff
diff --git a/examples/minimal/main.rs b/examples/minimal/main.rs
index d6b81cf..46009d7 100644
--- a/examples/minimal/main.rs
+++ b/examples/minimal/main.rs
@@ -17,6 +17,8 @@ use penrose::{
 };
 
 fn main() -> Result<()> {
+    simplelog::TermLogger::init(simplelog::LevelFilter::Trace, simplelog::Config::default(), simplelog::TerminalMode::Stdout).unwrap();
+
     let config = Config::default();
     let hooks = vec![];
 
diff --git a/src/xcb/api.rs b/src/xcb/api.rs
index bc7e449..de4ab17 100644
--- a/src/xcb/api.rs
+++ b/src/xcb/api.rs
@@ -169,6 +169,9 @@ impl Api {
             })
             .collect::<Result<HashMap<_, _>>>()?;
 
+        // cause an X11 error
+        self.mark_focused_window(0xdeadbeef);
+
         Ok(())
     }
 
@@ -261,12 +264,16 @@ impl Api {
     fn generic_xcb_to_xevent(&self, event: XcbGenericEvent) -> Result<Option<XEvent>> {
         let xcb_response_type_mask: u8 = 0x7F;
         let numlock = xcb::MOD_MASK_2 as u16;
+        println!("Got an event of type {}", event.response_type());
 
         let etype = event.response_type() & xcb_response_type_mask;
         // Need to apply the randr_base mask as well which doesn't seem to work in 'match'
         if etype == self.randr_base + xcb::randr::NOTIFY {
             return Ok(Some(XEvent::RandrNotify));
         }
+        if etype == self.randr_base + xcb::randr::SCREEN_CHANGE_NOTIFY {
+            return Ok(Some(XEvent::ScreenChange));
+        }
 
         Ok(match etype {
             xcb::BUTTON_PRESS | xcb::BUTTON_RELEASE | xcb::MOTION_NOTIFY => {
@@ -318,7 +325,8 @@ impl Api {
                 Some(XEvent::Destroy { id: e.window() })
             }
 
-            xcb::randr::SCREEN_CHANGE_NOTIFY => Some(XEvent::ScreenChange),
+            // This was 0 and thus causes an unreachable pattern warning below
+            //xcb::randr::SCREEN_CHANGE_NOTIFY => Some(XEvent::ScreenChange),
 
             xcb::CONFIGURE_NOTIFY => {
                 let e: &xcb::ConfigureNotifyEvent = unsafe { xcb::cast_event(&event) };
@@ -355,6 +363,7 @@ impl Api {
 
             xcb::PROPERTY_NOTIFY => {
                 let e: &xcb::PropertyNotifyEvent = unsafe { xcb::cast_event(&event) };
+        self.mark_focused_window(0xdeadbeef);
                 xcb::xproto::get_atom_name(&self.conn, e.atom())
                     .get_reply()
                     .ok()
@@ -373,6 +382,22 @@ impl Api {
                     })
             }
 
+            0 => {
+                let e: &xcb::GenericError = unsafe { xcb::cast_event(&event) };
+                // I am really not a fan of the xcb crate... :-(
+                let error = unsafe { *e.ptr };
+                println!("got an error");
+                error!(
+                    "X11 error: error seq {}, code {}, xid {}, request: {}:{}",
+                    error.sequence,
+                    error.error_code,
+                    error.resource_id,
+                    error.major_code,
+                    error.minor_code,
+                );
+                None
+            }
+
             // NOTE: ignoring other event types
             _ => None,
         })
```

**Comment by sminez** [comment]
The more I look into the `rust-xcb` crate (and hear your thoughts on it), the more I regret not spending longer looking for alternatives when I first started `penrose` :disappointed: 

It's always felt clunky and awkward but I thought that was down to me being a noob with X11 stuff and the docs being bad: not that the crate was straight up broken in multiple ways :grimacing:

**Comment by sminez** [comment]
This is going to be a fairly large change: if the `XConn` API is getting breaking changes then it makes sense to take a look at how it currently stands and tidy things up. This was one of the first parts of `penrose` to be written and it definitely shows...!

I'm currently working on reducing the methods needed for implementing `XConn` to the following (with default impls being provided for a lot of the existing methods in terms of some of the new ones):
```rust
trait XConn {
    #[cfg(feature = "serde")]
    fn hydrate(&mut self) -> Result<()>;
    fn init(&self) -> Result<()>;
    fn cleanup(&self) -> Result<()>;

    fn flush(&self) -> bool;
    fn wait_for_event(&self) -> Result<XEvent>;
    fn send_client_event(&self, id: Xid, atom_name: &str, data: &[u32]) -> Result<()>;
    fn grab_keys(
        &self,
        key_bindings: &KeyBindings<Self>,
        mouse_bindings: &MouseBindings<Self>,
    ) -> Result<()>
    where
        Self: Sized;

    fn current_screens(&self) -> Result<Vec<Screen>>;
    fn cursor_position(&self) -> Result<Point>;
    fn warp_cursor(&self, id: Option<Xid>, screen: &Screen) -> Result<()>;
    fn client_geometry(&self, id: Xid) -> Result<Region>;

    fn active_clients(&self) -> Result<Vec<Xid>>;
    fn focused_client(&self) -> Result<Xid>;

    fn map_client(&self, id: Xid) -> Result<()>;
    fn unmap_client(&self, id: Xid) -> Result<()>;
    fn focus_client(&self, id: Xid) -> Result<()>;
    fn list_props(&self, id: Xid) -> Result<Vec<String>>;
    fn get_prop(&self, id: Xid, name: &str) -> Result<Prop>;

    // -- new --
    fn root(&self) -> Xid;
    fn delete_prop(&self, id: Xid, name: &str) -> Result<()>;
    fn change_prop(&self, id: Xid, name: &str, val: Prop) -> Result<()>;
    fn configure_client(&self, id: Xid, data: &[ClientConfig]) -> Result<()>;
    fn set_client_attributes(&self, id: Xid, data: &[ClientAttr]) -> Result<()>;

    // -- Removed as it is only used in one place that can be worked around --
    // fn intern_atom(&self, atom: &str) -> Result<u32>;

    // -- Default implementations in terms of the above --
    fn window_should_float(&self, id: Xid, floating_classes: &[&str]) -> Result<bool>;
    fn is_managed_window(&self, id: Xid) -> Result<bool>;
    fn toggle_client_fullscreen(&self, id: Xid, client_is_fullscreen: bool) -> Result<()>;
    fn set_client_workspace(&self, id: Xid, wix: usize) -> Result<()>;
    fn set_root_window_name(&self, name: &str) -> Result<()>;
    fn set_current_workspace(&self, wix: usize) -> Result<()>;
    fn update_desktops(&self, workspaces: &[&str]) -> Result<()>;
    fn update_known_clients(&self, clients: &[Xid]) -> Result<()>;
    fn set_wm_properties(&self, check_win: Xid, workspaces: &[&str]) -> Result<()>;
    fn position_client(&self, id: Xid, r: Region, border: u32, stack_above: bool) -> Result<()>;
    fn set_client_border_color(&self, id: Xid, color: Color) -> Result<()>;
    fn mark_new_client(&self, id: Xid) -> Result<()>;
    fn raise_client(&self, id: Xid) -> Result<()>;
}
```

`WinConfig` and `WinAttr` are being renamed to `ClientConfig` and `ClientAttr` respectively to tidy up the existing confusion around Window vs Client naming conventions. They are also being moved to the `xconnection` module as this is the only place they are used in core. (They are of course, then used in implementations of `XConn` as well).

`data_types::PropVal` is being removed in favour of just using `xconnection::Prop` which works with the new code around querying for and setting client properties correctly.

I'm also going to rename `WinId` to `Xid` given that all X resources use a `u32` for their ID. `WinId` will be kept as an alias to avoid breaking existing public apis where users may have specified this explicitly.

**Comment by sminez** [comment]
Implemented in #130
